engine/src/main/java/org/terasology/entitySystem/entity/internal/PojoEntityCache.java

```diff
@@ -0,0 +1,350 @@
+import org.terasology.entitySystem.prefab.Prefab;
+import org.terasology.logic.location.LocationComponent;
+import org.terasology.math.geom.Quat4f;
+import org.terasology.math.geom.Vector3f;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+import static org.terasology.entitySystem.entity.internal.PojoEntityManager.NULL_ID;
+
+/**
+ */
+public class PojoEntityCache implements EntityCache {
+
+    private PojoEntityManager entityManager;
+
+    private static final Logger logger = LoggerFactory.getLogger(PojoEntityCache.class);
+
+    private Map<Long, BaseEntityRef> entityStore = new MapMaker().weakValues().concurrencyLevel(4).initialCapacity(1000).makeMap();
+    private ComponentTable componentStore = new ComponentTable();
+
+    public PojoEntityCache(PojoEntityManager entityManager) {
+        this.entityManager = entityManager;
+    }
+
+    @Override
+    public void clear() {
+        //Todo: should also clear out ids from the EntityManager
+        entityStore.values().forEach(BaseEntityRef::invalidate);
+        componentStore.clear();
+        entityStore.clear();
+    }
+
+
+    @Override
+    public EntityRef create() {
+        EntityRef entityRef = createEntityRef(entityManager.createEntity());
+        /*
+         * The entity change listener are also used to detect new entities. By adding one component we inform those
+         * listeners about the new entity.
+         */
+        entityRef.addComponent(new EntityInfoComponent());
+        return entityRef;
+    }
+
+    @Override
+    public EntityRef create(Component... components) {
+        return create(Arrays.asList(components));
+    }
+
+    @Override
+    public EntityRef create(Iterable<Component> components) {
+        EntityRef entity = createEntity(components);
+
+        EventSystem eventSystem = entityManager.getEventSystem();
+        if (eventSystem != null) {
+            eventSystem.send(entity, OnAddedComponent.newInstance());
+            eventSystem.send(entity, OnActivatedComponent.newInstance());
+        }
+        for (Component component: components) {
+            entityManager.notifyComponentAdded(entity, component.getClass());
+        }
+        return entity;
+    }
+
+    @Override
+    public EntityRef create(String prefabName) {
+        if (prefabName != null && !prefabName.isEmpty()) {
+            Prefab prefab = entityManager.getPrefabManager().getPrefab(prefabName);
+            if (prefab == null) {
+                logger.warn("Unable to instantiate unknown prefab: \"{}\"", prefabName);
+                return EntityRef.NULL;
+            }
+            return create(prefab);
+        }
+        return create();
+    }
+
+    @Override
+    public EntityRef create(String prefabName, Vector3f position) {
+        if (prefabName != null && !prefabName.isEmpty()) {
+            Prefab prefab = entityManager.getPrefabManager().getPrefab(prefabName);
```

Comment:
Ah, I'll add the checks where they're needed.

Some of them are mostly safe because the associated create(String prefabName) does the null check, but I'll add the checks there as well, in case it is called with a null prefab from somewhere else (currently nowhere within the engine calls the create() methods with a Prefab directly, but it doesn't hurt to be safe).